### DISCUSSION TITLE
Resizing image after resampling voxels to unit isotropy

### DISCUSSION THREAD

**Discussion by mattwarkentin** [TorchIO-project/torchio#690] [General]
Hi @fepegar,

I just had a quick question that follows from some of the discussions we had previously about how `Resize()` may deform the aspect ratio if not used carefully.

If you first resampled the volume to unit isotropic voxel spacing (i.e. 1mm x 1mm x 1mm), and then resized to some shape, would this avoid the deformation? Seems like it would, but sometimes my mental model fails me.

Something like...

```py
img = tio.ScalarImage(...)

trans = tio.Compose([
  tio.ToCanonical(),
  tio.Resample(1),
  tio.Clamp(-1000, 1000),
  tio.Resize(200)
])

img_t = trans(img)
```

**Comment by fepegar** [comment]
The problem here is that resizing _is_ resampling. What it does is change the voxel spacing so that the field of view remains constant while the number of voxels along each dimension is modified. The new spacing is a function of the original shape, therefore your resampling to 1 mm iso is wasted if your images have different shapes. If you look at the images in a smart viewer like 3D Slicer, they will look very similar because they remain in the same physical space. But for a CNN, which doesn't care about voxel spacing, one image will look rescaled with respect to the other.

Does this make sense? I can write a short demo if it doesn't.

**Reply by mattwarkentin** [reply]
Hmm, yes, I think I am beginning to understand more - though I would never say no to a short demo that helps improve my understanding.

I guess what matters to me is that:
- I have images that have different spacing and different spatial shape, and I need them to be the same
- The inputs to the CNN need to all be the same shape, but spacing is ignored, so this is achievable safely with `Resample(1)` + `CropOrPad(...)`, however, this means some images will have lots of padding, while some may actually be cropping into a relevant/important area for learning. You could `CropOrPad` all of the volumes to be the size of the largest volume, so that none of the volumes get cropped at all, but this leads to an increase in the sizes of tensors/memory consumption and higher resolution images than are probably needed for the learning task.

Resizing has the "nice" feature that it maintains the FOV, but I suppose since it distorts the aspect ratio/scale, this would be problematic.

Any thoughts?

**Reply by fepegar** [reply]
Here's a notebook that shows how the image, once spacing is ignored, is deformed: https://gist.github.com/fepegar/5d9e66d8302a110975efa932cd1ab2a0

If you can share some info about your dataset, I'm happy to suggest preprocessing ideas if I can come up with any.
